### PR TITLE
[#28] Email verification

### PR DESCRIPTION
**Pull request by jswk**
Add token based email verification. An AppUser is created with field
enabled=false, and it can be activated using the token sent by mail.

The logic of sending emails is decoupled from the controller to async
event listeners.

Create LoggingMailSender, which outputs email contents to log.

For testing I introduced GreenMail, which is used to verify appropriate
emails are sent.

The email contents are put in messages_pl_PL.properties.

### CODE REVIEW

**Comment by michal-szostak** [comment]
I updated issue scope with new requirements: #28

**Comment by jswk** [comment]
Another update in #28.

**Comment by jswk** [comment]
I updated the code to match the updated requirements. @michal-szostak please verify if it works as expected.

File: s4e-backend/src/main/java/pl/cyfronet/s4e/bean/RegistrationToken.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+import javax.persistence.*;
+import javax.validation.constraints.NotEmpty;
+import javax.validation.constraints.NotNull;
+import java.time.LocalDateTime;
+
+@Entity
+@Data
+@Builder
+public class RegistrationToken {
```
Comment (kmarszalek) [inline s4e-backend/src/main/java/pl/cyfronet/s4e/bean/RegistrationToken.java]: not emailverification?

Comment (jswk) [reply s4e-backend/src/main/java/pl/cyfronet/s4e/bean/RegistrationToken.java]: I was considering EmailVerificationToken and it was too long for me. But actually just EmailVerification sounds better. The token is a component of such verification then. I'll refactor it.

File: s4e-backend/src/main/java/pl/cyfronet/s4e/service/AppUserService.java [hunk 2]
```diff
@@ -24,4 +26,8 @@ public AppUser save(AppUser appUser) throws AppUserCreationException {
             throw new AppUserCreationException(e);
         }
     }
+
+    public Optional<AppUser> findByEmail(String email) {
```
Comment (kmarszalek) [inline s4e-backend/src/main/java/pl/cyfronet/s4e/service/AppUserService.java]: should we return from services one type of return, optional or e.g. AppUser/boolean ?

File: s4e-backend/src/main/java/pl/cyfronet/s4e/service/VerificationTokenService.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+import java.time.Duration;
+import java.time.LocalDateTime;
+import java.time.temporal.TemporalAmount;
+import java.util.Optional;
+import java.util.UUID;
+
+@Service
+@RequiredArgsConstructor
+public class VerificationTokenService {
```
Comment (kmarszalek) [inline s4e-backend/src/main/java/pl/cyfronet/s4e/service/VerificationTokenService.java]: same as other service :)

File: s4e-backend/src/main/java/pl/cyfronet/s4e/controller/AppUserController.java [hunk 4]
```diff
@@ -22,20 +32,87 @@
...
+                    @ExampleProperty(mediaType = "application/json", value = "{\"email\":[\"musi być adresem e-mail\"]}")
+            }))
+    })
     @PostMapping("/register")
-    public ResponseEntity<?> register(@RequestBody @Valid RegisterRequest registerRequest) throws AppUserCreationException {
-        appUserService.save(AppUser.builder()
+    public ResponseEntity<?> register(
+            @RequestBody @Valid RegisterRequest registerRequest
+    ) throws AppUserCreationException {
+        AppUser appUser = appUserService.save(AppUser.builder()
```
Comment (kmarszalek) [inline s4e-backend/src/main/java/pl/cyfronet/s4e/controller/AppUserController.java]: check if exists? repo.save do save or update...

**Pull request merged**